aptos-move/mvhashmap/src/lib.rs

```diff
@@ -1,304 +1,120 @@
-        // TODO: investigate logical deletion.
-        let mut v = self.data.get_mut(key).expect("Path must exist");
-        v.versioned_map.remove(&txn_idx);
+        match key.module_path() {
+            Some(_) => self.code.delete(key, txn_idx),
+            None => self.data.delete(key, txn_idx),
+        }
     }
 
-    /// Read entry from transaction 'txn_idx' at access path 'key'.
-    pub fn read(
-        &self,
-        key: &K,
-        txn_idx: TxnIndex,
-    ) -> anyhow::Result<MVHashMapOutput<V>, MVHashMapError> {
-        use MVHashMapError::*;
-        use MVHashMapOutput::*;
-
-        match self.data.get(key) {
-            Some(v) => {
-                let mut iter = v.versioned_map.range(0..txn_idx);
-
-                // If read encounters a delta, it must traverse the block of transactions
-                // (top-down) until it encounters a write or reaches the end of the block.
-                // During traversal, all aggregator deltas have to be accumulated together.
-                let mut accumulator: Option<Result<DeltaOp, ()>> = None;
-                while let Some((idx, entry)) = iter.next_back() {
-                    let flag = entry.flag();
+    /// Add a versioned write at a specified key, in code or data map according to the key.
+    pub fn write(&self, key: &K, version: Version, value: V) {
+        match key.module_path() {
+            Some(_) => self.code.write(key, version.0, value),
+            None => self.data.write(key, version, value),
+        }
+    }
 
-                    if flag == FLAG_ESTIMATE {
-                        // Found a dependency.
-                        return Err(Dependency(*idx));
-                    }
+    // -----------------------------------------------
+    // Functions specific to the multi-versioned data.
 
-                    // The entry should be populated.
-                    debug_assert!(flag == FLAG_DONE);
+    /// Add a delta at a specified key.
+    pub fn add_delta(&self, key: &K, txn_idx: usize, delta: DeltaOp) {
+        self.data.add_delta(key, txn_idx, delta);
+    }
 
-                    match (&entry.cell, accumulator.as_mut()) {
-                        (EntryCell::Write(incarnation, data), None) => {
-                            // Resolve to the write if no deltas were applied in between.
-                            let write_version = (*idx, *incarnation);
-                            return Ok(Version(write_version, data.clone()));
-                        },
-                        (EntryCell::Write(incarnation, data), Some(accumulator)) => {
-                            // Deltas were applied. We must deserialize the value
-                            // of the write and apply the aggregated delta accumulator.
+    /// Read data at access path 'key', from the perspective of transaction 'txn_idx'.
+    pub fn read(&self, key: &K, txn_idx: TxnIndex) -> anyhow::Result<MVDataOutput<V>, MVDataError> {
+        self.data.read(key, txn_idx)
```

Comment:
I'll change to read_data, good point.
Unbelievable, I was deliberating between dispatching in read for now or using this placeholder-like fetch_module, but somehow I didn't include either? Why does the forge pass? Oh because we never need to read published modules and it comes from storage... LoL I'd have broken everything. Let me dispatch on view's read for now, and add a test that makes sure fetch module returns the correct blob.

Later fetch module will either return blob or executable (if already recorded), and record for validation only when it returns the executable - so VM will store it once compiled - except init_module which actually should just work because it's staged inside the VM's data_store (just need to add Option<Executable> there as well).

Reply:
oh, LOL, the read write fallback saves it. If a module got published, and we try to read it we anyway fallback to sequential right now. If it didn't get published, the read will not find it which is a correct behavior and will come from storage = )

I should still probably fix it, right? :D but it's hard to write a PE test, can just test on MVHashMap side

Reply:
@zekun000 actually maybe this is a good plan for rolling out:
- we keep the module read-write fallback even after adding the new fetch & validation flow for the modules. This way, the current VM will keep using the get_state_view to get the module data, and even while landing the new code to aptos-core, things will work as normal.
- we then make changes to Move-VM to start using the new APIs and manage modules/structs/functions/links a bit different, and this is the change that we will test, so it will basically be a versioning of VM new vs old. Once we are confident in new we switch to it fully.
- somewhere here we may want to introduce changes to aptos-vm for testing the new version across the blocks - hopefully just a simple way to know if we are using the new flow, do a few more things it can support to test properly
- finally, we can get rid of the fallback logic from the parallel executor, get rid of the get_state_view interface (have dedicated ones for data / modules), and get rid of all the flushing.

@runtian-zhou @vgao1996 as well.

Reply:
I think keep the read-write fallback is fine, but we probably still want to keep expected behavior on interfaces i.e. read versioned data instead of going to storage and rely on fallbacks?

Reply:
makes sense - there will be totally new interfaces for modules, but I'll patch the old common interface (for data and code)`get_state_view` to dispatch and return the right versioned blob if it exists. and it may exist because the r/w intersection is checked right after.
maybe we can even use it to assert somehow but not super sure, will look and update the PR today.

Reply:
added things to view.rs, ugly, but temporary